pkg/operator/ceph/nvmeof/spec.go

```diff
@@ -0,0 +1,451 @@
+	return nil
+}
+
+func (r *ReconcileCephNVMeOFGateway) makeDeployment(nvmeof *cephv1.CephNVMeOFGateway, daemonID, configHash string) (*apps.Deployment, error) {
+	resourceName := instanceName(nvmeof, daemonID)
+
+	deployment := &apps.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      resourceName,
+			Namespace: nvmeof.Namespace,
+			Labels:    getLabels(nvmeof, daemonID, true),
+		},
+	}
+
+	hostNetwork := nvmeof.IsHostNetwork(r.cephClusterSpec)
+
+	k8sutil.AddRookVersionLabelToDeployment(deployment)
+	controller.AddCephVersionLabelToDeployment(r.clusterInfo.CephVersion, deployment)
+	nvmeof.Spec.Annotations.ApplyToObjectMeta(&deployment.ObjectMeta)
+	nvmeof.Spec.Labels.ApplyToObjectMeta(&deployment.ObjectMeta)
+
+	cephConfigVol, cephConfigMount := cephConfigVolumeAndMount()
+	gatewayConfigVol, _ := gatewayConfigVolumeAndMount(nvmeof)
+
+	podSpec := v1.PodSpec{
+		InitContainers: []v1.Container{
+			r.connectionConfigInitContainer(nvmeof),
+		},
+		Containers: []v1.Container{
+			r.daemonContainer(nvmeof, cephConfigMount),
+		},
+		RestartPolicy: v1.RestartPolicyAlways,
+		Volumes: []v1.Volume{
+			cephConfigVol,
+			adminKeyringVolume(),
+			gatewayConfigVol,
+		},
+		HostNetwork:        hostNetwork,
+		PriorityClassName:  nvmeof.Spec.PriorityClassName,
+		SecurityContext:    &v1.PodSecurityContext{},
+		ServiceAccountName: serviceAccountName,
+	}
+
+	k8sutil.AddUnreachableNodeToleration(&podSpec)
+
+	if hostNetwork {
+		podSpec.DNSPolicy = v1.DNSClusterFirstWithHostNet
+	}
+	nvmeof.Spec.Placement.ApplyToPodSpec(&podSpec)
+
+	podTemplateSpec := v1.PodTemplateSpec{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   resourceName,
+			Labels: getLabels(nvmeof, daemonID, true),
+			Annotations: map[string]string{
+				"config-hash": configHash,
+			},
+		},
+		Spec: podSpec,
+	}
+
+	if hostNetwork {
+		podSpec.DNSPolicy = v1.DNSClusterFirstWithHostNet
+	} else if r.cephClusterSpec.Network.IsMultus() {
+		if err := k8sutil.ApplyMultus(r.clusterInfo.Namespace, &r.cephClusterSpec.Network, &podTemplateSpec.ObjectMeta); err != nil {
+			return nil, err
+		}
+	}
+
+	nvmeof.Spec.Annotations.ApplyToObjectMeta(&podTemplateSpec.ObjectMeta)
+	nvmeof.Spec.Labels.ApplyToObjectMeta(&podTemplateSpec.ObjectMeta)
+
+	replicas := int32(1)
+	deployment.Spec = apps.DeploymentSpec{
+		RevisionHistoryLimit: controller.RevisionHistoryLimit(),
+		Selector:             &metav1.LabelSelector{MatchLabels: getLabels(nvmeof, daemonID, false)},
+		Template:             podTemplateSpec,
+		Replicas:             &replicas,
+	}
+
+	return deployment, nil
+}
+
+func (r *ReconcileCephNVMeOFGateway) connectionConfigInitContainer(nvmeof *cephv1.CephNVMeOFGateway) v1.Container {
+	_, cephConfigMount := cephConfigVolumeAndMount()
+	_, gatewayConfigMount := gatewayConfigVolumeAndMount(nvmeof)
+
+	poolName := nvmeof.Spec.Pool
+	if poolName == "" {
+		poolName = "nvmeofpool"
+	}
+
+	script := fmt.Sprintf(`
+set -xEeuo pipefail
```

Comment:
I'd suggest that we not inline this script in the go code. Instead, I've found development and maintenance to be easier with separate scripts that are loaded into strings. We do this with the MDS liveness probe: 

https://github.com/rook/rook/blob/master/pkg/operator/ceph/file/mds/livenessprobe.sh

https://github.com/rook/rook/blob/master/pkg/operator/ceph/file/mds/livenessprobe.go